sysboundary/ionosphere.h

```diff
@@ -148,20 +151,130 @@ namespace SBC {
+         {          0,            0,            0,            0},
+         {          0,            0,            0,            0}
+      }};
+      
+      /*!< Parametrization coefficients Ci for electrons and protons, the species is included in the i index.
+       * \param i index into P_ij
+       * \param E energy in keV
+       */
+      inline Real fC(cint i, creal E) {
+         return exp(P_ij[i][0] + P_ij[i][1]*log(E) + P_ij[i][2]*log(E)*log(E) + P_ij[i][3]*log(E)*log(E)*log(E));
+      }
+      
+      /*!< normalised column mass y as a function of altitude h and energy E, for
+       * - electrons Fang 2010 https://doi.org/10.1029/2010GL045406 equation (1)
+       * - protons Fang 2013 https://doi.org/10.1002/jgra.50484 equation (5)
+       * 
+       * / 1000 for kg m^-3 to g cm^-3
+       * * 100 for m to cm
+       * results in / 10 inside pow().
+       * 
+       * \param p particle species enum Particles::PROTON and ELECTRON
+       * \param E energy in keV
+       * \param h altitude index
+       */
+      inline Real fangNormalizedColumnMass(cint p, creal E, cuint h) {
+         Real y;
+         switch(p) {
+            case Particles::PROTON:
+               y = 7.5 / E * pow(atmosphere[h].density * atmosphere[h].scaleHeight / 1e-4 / 10, 0.9);
+               break;
+            case Particles::ELECTRON:
+               y = 2 / E * pow(atmosphere[h].density * atmosphere[h].scaleHeight / 6e-6 / 10, 0.7);
+               break;
+            default:
+               cerr << "Invalid precipitating species!" << endl;
+               abort();
+         }
+         return y;
+      }
+      
+      /*!< Energy dissipation f as a function of altitude and energy E, for
+       * - electrons Fang 2010 https://doi.org/10.1029/2010GL045406 equation (4)
+       * - protons Fang 2013 https://doi.org/10.1002/jgra.50484 equation (6)
+       * 
+       * Note that the i indices are taken -1 as we are 0-indexing in the arrays P_ij.
+       * 
+       * \param p particle species enum Particles::PROTON and ELECTRON
+       * \param E energy in keV
+       * \param h altitude index
+       */
+      inline Real fangEnergyDissipation(cint p, creal E, cint h) {
+         creal y = fangNormalizedColumnMass(p, E, h);
+         
+         cint offset = p*12; // 0 for proton, 12 for electron, see indexing in P_ij
+         return
+           fC(offset+0,E) * pow(y, fC(offset+1,E)) * exp(-fC(offset+ 2,E) * pow(y, fC(offset+ 3,E)))
+         + fC(offset+4,E) * pow(y, fC(offset+5,E)) * exp(-fC(offset+ 6,E) * pow(y, fC(offset+ 7,E)))
+         + fC(offset+8,E) * pow(y, fC(offset+9,E)) * exp(-fC(offset+10,E) * pow(y, fC(offset+11,E))); // this line returns zero for electrons, see P_ij
+      }
+      
+      Real totalIonizationFang(
+         cint n,
+         cint h
+      );
       
       // Hardcoded constants for calculating ion production table
       // TODO: Make these parameters?
```

Comment:
Just a note here, to avoid mixing/breaking things I kept the lookup table parameters from the Rees/ Sergienko & Ivanov paths separate from the Fang one so far. We can unify and indeed then make them all cfg parameters later. Actually I think that the energy ranges shouldn't be touched for the Fang models as they are now by default the stated validity range of the parametrisations.

Reply:
I guess it would simplify the code a lot if the other models, too, were to get rid of the precomputed lookup table.
We would need to measure the performance impact of that though (I'd wager it is not large)